### PR TITLE
fix: Unit tests

### PR DESCRIPTION
**Pull request by ajhelsby**
## This PR
Improve unit test coverage to ensure spec is followed.

### CODE REVIEW

File: tests/test_open_feature_client.py [hunk 1]
```diff
@@ -0,0 +1,173 @@
...
+def setup():
+    api.set_provider(NoOpProvider())
+    provider = api.get_provider()
+    assert isinstance(provider, NoOpProvider)
+
+
+def test_should_use_no_op_provider_if_none_provided():
+    # Given
+    # When
+    flag = OpenFeatureClient("No provider", "1.0")._create_provider_evaluation(
```
[the remark this answers is not in the document]
Comment (ajhelsby) [reply tests/test_open_feature_client.py:24]: Avoiding having every single type (bool, string etc) having to test this functionality when they're all using the same method

[the remark this answers is not in the document]
Comment (tcarrio) [reply tests/test_open_feature_client.py:24]: This sounds like the perfect case for parameterized test cases. You can write a single method that is reused for a list of inputs.

The goal will be to have the client fully covered with unit tests, including each individual method, and it wouldn't require as much code to do it that way.

File: tests/test_open_feature_client.py [hunk 2]
```diff
@@ -0,0 +1,173 @@
...
+    # Then
+    assert ge.value
+    assert ge.value.error_message == "Unknown flag type"
+    assert ge.value.error_code == ErrorCode.GENERAL
+
+
+def test_should_get_boolean_flag(no_op_provider_client):
+    # Given
+    # When
+    flag = no_op_provider_client.get_boolean_details(flag_key="Key", default_value=True)
```
[the remark this answers is not in the document]
Comment (tcarrio) [reply tests/test_open_feature_client.py:49]: My suggestion would be to create a mock provider with `MagicMock` here  that's shared, similar to the unit tests in the `hooks/conftest.py`

File: tests/conftest.py [hunk 3]
```diff
@@ -0,0 +1,55 @@
...
+        pass
+
+    def finally_after(self, hook_context: HookContext, hints: dict):
+        pass
+
+    def supports_flag_value_type(self, flag_type: FlagType) -> bool:
+        return True
+
+
+class TestOpenFeatureErrorHook(Hook):
```
Comment (tcarrio) [inline tests/conftest.py:39]: This is a case where mocking would be more useful- you can also validate that the provider is called with the correct values using mock assertions like

`mock_provider.before.asset_called_with(...)`

Comment (tcarrio) [reply tests/conftest.py:39]: Another option would be to use a spy, then you can have shared predefined logic with a real class but you would spy on it to run those assertions.

File: tests/test_open_feature_client.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (ajhelsby) [reply tests/test_open_feature_client.py]: 👍🏻 Done

File: tests/test_open_feature_api.py [hunk 4]
```diff
@@ -0,0 +1,56 @@
...
+    assert (
+        ge.value.error_message
+        == "Provider not set. Call set_provider before using get_client"
+    )
+    assert ge.value.error_code == ErrorCode.GENERAL
+
+
+def test_should_return_open_feature_client_when_configured_correctly():
+    # Given
+    set_provider(NoOpProvider())
```
[the remark this answers is not in the document]
Comment (ajhelsby) [reply tests/test_open_feature_api.py:25]: This has now been fixed through a fixture called `clear_provider`

File: tests/conftest.py (hunk 3 above)
Comment (ajhelsby) [reply tests/conftest.py]: The hooks are now mocked

File: tests/test_open_feature_client.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (ajhelsby) [reply tests/test_open_feature_client.py]: Hooks are now mocked

File: tests/conftest.py [hunk 5]
```diff
@@ -0,0 +1,62 @@
...
+from open_feature.flag_evaluation.flag_type import FlagType
+from open_feature.hooks.hook import Hook
+from open_feature.hooks.hook_context import HookContext
+from open_feature.provider.no_op_provider import NoOpProvider
+
+
+@pytest.fixture(autouse=True)
+def clear_provider():
+    yield
+    # Setting private provider back to None to ensure other tests aren't affected
```
[the remark this answers is not in the document]
Comment (ajhelsby) [reply tests/conftest.py:10]: Updated

**Review by matthewelwell — APPROVED** [review]

**Pull request merged**